base/src/main/java/org/aya/terck/CallGraph.java

```diff
@@ -2,31 +2,90 @@
 // Use of this source code is governed by the MIT license that can be found in the LICENSE.md file.
 package org.aya.terck;
 
+import kala.collection.immutable.ImmutableSeq;
+import kala.collection.mutable.MutableLinkedHashMap;
 import kala.collection.mutable.MutableMap;
 import kala.collection.mutable.MutableSet;
+import kala.control.Option;
+import kala.value.Ref;
 import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
 
-public record CallGraph<T>(
-  @NotNull MutableMap<T, @NotNull MutableMap<T, MutableSet<@NotNull CallMatrix<T>>>> graph
+/**
+ * Call graph is a multi-graph; each vertex represents a definition and each edge from vertex `f`
+ * to vertex `g` represents a call to `g` within `f`. The edges are labeled with call matrices,
+ * and can be labelled with several call matrices if there are several paths from `f` to `g`.
+ *
+ * @author kiva
+ * @see CallMatrix
+ */
+public record CallGraph<T, P>(
+  @NotNull MutableMap<T, @NotNull MutableMap<T, MutableSet<@NotNull CallMatrix<T, P>>>> graph
 ) {
-  public static <T> @NotNull CallGraph<T> create() {
-    return new CallGraph<>(MutableMap.create());
+  public static <T, P> @NotNull CallGraph<T, P> create() {
+    return new CallGraph<>(MutableLinkedHashMap.of());
   }
 
   /**
-   * Add a call matrix to the graph, if there's no same call matrix from
-   * {@link CallMatrix#caller()} to {@link CallMatrix#callee()} was found.
+   * Add a call matrix to the graph, if there's no smaller call matrix from
+   * {@link CallMatrix#domain()} to {@link CallMatrix#codomain()} was found.
    *
    * @return whether the matrix it is added to the graph
    */
-  public boolean put(@NotNull CallMatrix<T> matrix) {
-    var caller = matrix.caller();
-    var callee = matrix.callee();
-    var set = graph.getOrPut(caller, MutableMap::create)
+  public boolean put(@NotNull CallMatrix<T, P> matrix) {
+    var caller = matrix.domain();
+    var callee = matrix.codomain();
+    var set = graph.getOrPut(caller, MutableLinkedHashMap::of)
       .getOrPut(callee, MutableSet::create);
-    // TODO: check if there's already a smaller call matrix?
     if (set.contains(matrix)) return false;
+    var unknown = set.anyMatch(arrow -> arrow.compare(matrix) != Relation.Unknown);
+    if (unknown) return false; // got stuck, fail early
+    set.removeAll(existing -> matrix.compare(existing) == Relation.LessThan);
     set.add(matrix);
     return true;
   }
+
+  /** the completion of a call graph is finding its transitive closure */
+  private static <T, P> @NotNull CallGraph<T, P> complete(@NotNull CallGraph<T, P> start) {
+    var oldGraph = new Ref<>(start);
+    var newEdge = new Ref<>(1);
+    // TODO: do we really need this fuel since we are just looking for a fixpoint?
```

Comment:
What does Agda/Arend do?

Reply:
They don't have it. I think the fuel can reduce unnecessary loops because we only need to traverse each edge in this graph (according to Warshall’s algorithm)